### ISSUE TITLE
Plot user provided data

### ISSUE THREAD

**Issue by c410-f3r** [bheisler/criterion.rs#198]
It would be nice to plot user provided data for those who care about the generated results of each benchmark evaluation. E.g.: Given an array of N points, each point could represent a diagonal of a [scatter-plot matrix](http://bastian.rieck.ru/blog/posts/2017/scatterplot_matrices_gnuplot/).

If possible, I will try to implement this feature myself. What do you think, @bheisler?

**Comment by bheisler** [comment]
Hey, thanks for trying Criterion.rs, and thanks for the suggestion.

I don't think I understand the use case for something like this. If you want to plot your own data, just use a plotting library, no?

**Comment by c410-f3r** [comment]
Yes! There are some native plotting libraries in Rust but would be nice to plot arbitrary user data alongside `Criterion` reports. By leveraging the power of the existing structure of this crate, users could measure the performance and the **quality** of their algorithms instead of creating a standalone benchmark/plotting suite, thus, reinventing the wheel.

I can close this issue if you think that this feature is out of scope of `Criterion` but I could at least give it a try. There are many fields that care about the their final results, like: Machine learning, Linear algebra, Linear programming, Dynamic programming and Heuristics.

**Comment by bheisler** [comment]
I see what you're saying - There are a couple of wish-list items that might cover this. I've never gotten around to implementing them. It's a lot of new API surface that needs to be defined and then maintained. Also, the more configuration options I expose now, the more things will break when I eventually change Criterion.rs to use the new test-framework system. If we're going to implement something like this, I think it should be locked behind a Cargo feature flag for now, to leave some flexibility to iterate on the interface.

I'm not really clear on what you're looking for here. Are you looking to inject some extra HTML into the generated reports (which might include custom chart images) or maybe even entirely replace the HTML reports? Alternately, are you looking to augment or replace the measurement code so that you can measure things other than runtime (eg. prediction accuracy)?

**Comment by c410-f3r** [comment]
The cargo features looks like a good idea and I would like to start with something more conservative, i.e., add an HTML section just to illustrate a generated scatter-plot matrix that is generic enough to handle N dimensions.

To get beyond this, we would have to care about the various types of data formats the user could give or how this data could be analyzed to provide additional info. All of this can be discussed afterwards.

the cargo features looks like a good idea.

**Comment by c410-f3r** [comment]
Development was moved to a new repository (https://github.com/criterion-rs/criterion.rs)
